File: test/rail_verify_public_test.py
"""
Public, self-contained Python rail-verification demo — runnable by anyone who clones the verifier.
Loads ONLY rail-verify.py (flat, via importlib since the filename is hyphenated); uses the SAME static
pre-signed fixtures (rail-fixtures.json) and CANNED rail responses — no live network, no emit-path
builder, no monorepo deps. Deps: rfc8785 + eth-account (requirements.txt).

Run:  pip install -r requirements.txt  &&  python3 test/rail_verify_public_test.py
"""
import importlib.util
import json
import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
spec = importlib.util.spec_from_file_location("rail_verify", os.path.join(HERE, "..", "rail-verify.py"))
rv = importlib.util.module_from_spec(spec)
spec.loader.exec_module(rv)

with open(os.path.join(HERE, "rail-fixtures.json")) as f:
    FIXTURES = json.load(f)


def to_read(receipt, rail_input):
    if rail_input.get("chain") == "xrpl":
        return rv.parse_xrpl_settlement(receipt, rail_input["legResults"])
    if rail_input.get("chain") == "solana":
        return rv.parse_solana_settlement(receipt, rail_input["getTx"])
    return rail_input  # already a RailRead


def main():
    failures = 0
    for fx in FIXTURES:
        a = rv.assess_rail_settlement(fx["receipt"], to_read(fx["receipt"], fx["railInput"]))
        exp = fx["expected"]
        checks = [
            ("railVerdict", a["railVerdict"], exp["railVerdict"]),
            ("attestation", a["attestation"], exp["attestation"]),
            ("protocolEnforced1pct", a["protocolEnforced1pct"], exp["protocolEnforced1pct"]),
        ]
        ok = all(got == want for _, got, want in checks)
        # honesty invariants (every verdict)
        ok = ok and ("tamper-evident, NOT omission-evident" in a["note"])
        if fx["receipt"]["enforcement"] == "attested_off_ledger":
            ok = ok and (a["protocolEnforced1pct"] is False)
        if ok:
            print(f"  OK  {fx['name']}: {a['railVerdict']} / {a['attestation']} / protocolEnforced1pct={a['protocolEnforced1pct']}")
        else:
            failures += 1
            print(f"  FAIL {fx['name']}: got {[ (n, g) for n, g, w in checks ]} expected {exp}")
    if failures:
        print(f"\n{failures} FAILED")
        sys.exit(1)
    print(f"\nall {len(FIXTURES)} public rail fixtures passed (Python)")


if __name__ == "__main__":
    main()

File: test/rail-verify.public.test.ts
// Public, self-contained rail-verification demo test — runnable by anyone who clones the verifier.
// Imports ONLY ./rail-verify (flat layout); uses STATIC pre-signed receipt fixtures (rail-fixtures.json)
// and CANNED rail RPC responses — no live network, no emit-path builder, no monorepo deps, no ethers.
// Mirrors the existing test/rejection.test.ts style.
//
// Run:  node --import tsx --test test/rail-verify.public.test.ts   (deps: tsx + viem only)
//
// "Verify it yourself" demonstrations of the 3-rail per-rail settlement confirm + its honesty rules.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import {
  assessRailSettlement, parseXrplSettlement, parseSolanaSettlement,
  type SettlementReceipt, type RailRead,
} from "../rail-verify";

type Fixture = {
  name: string;
  receipt: SettlementReceipt;
  railInput: RailRead | { chain: "xrpl"; legResults: Array<{ reachable: boolean; tx: unknown }> } | { chain: "solana"; getTx: { reachable: boolean; tx: unknown } };
  expected: { railVerdict: string; attestation: string; protocolEnforced1pct: boolean };
};

const FIXTURES: Fixture[] = JSON.parse(readFileSync(new URL("./rail-fixtures.json", import.meta.url), "utf8"));

// resolve a fixture's railInput into a RailRead: chain-raw inputs run through the REAL parser
// (parseXrplSettlement / parseSolanaSettlement); a ready RailRead is used directly.
function toRead(receipt: SettlementReceipt, railInput: Fixture["railInput"]): RailRead {
  if ("chain" in railInput && railInput.chain === "xrpl") return parseXrplSettlement(receipt, railInput.legResults as any);
  if ("chain" in railInput && railInput.chain === "solana") return parseSolanaSettlement(receipt, railInput.getTx as any);
  return railInput as RailRead;
}

describe("rail-verify — public 3-rail settlement confirm demos (static fixtures, offline)", () => {
  for (const fx of FIXTURES) {
    it(fx.name, async () => {
      const a = await assessRailSettlement(fx.receipt, toRead(fx.receipt, fx.railInput));
      assert.equal(a.railVerdict, fx.expected.railVerdict, `railVerdict for ${fx.name}`);
      assert.equal(a.attestation, fx.expected.attestation, `attestation for ${fx.name}`);
      assert.equal(a.protocolEnforced1pct, fx.expected.protocolEnforced1pct, `protocolEnforced1pct for ${fx.name}`);
      // honesty invariants that hold for EVERY verdict:
      assert.match(a.note, /tamper-evident, NOT omission-evident/);             // omission caveat always present
      if (fx.receipt.enforcement === "attested_off_ledger") assert.equal(a.protocolEnforced1pct, false); // off-ledger never enforced
      console.log(`  ✓ ${fx.name}: ${a.railVerdict} / ${a.attestation} / protocolEnforced1pct=${a.protocolEnforced1pct}`);
    });
  }
});
